src/validation.cpp

```diff
@@ -5139,3 +5147,82 @@ class CMainCleanup
     }
 };
 static CMainCleanup instance_of_cmaincleanup;
+
+uint256 ChainstateManager::SnapshotBlockhash() const {
+    if (m_active_chainstate != nullptr) {
+        // If a snapshot chainstate exists, it will always be our active.
+        return m_active_chainstate->m_from_snapshot_blockhash;
+    }
+    return {};
+}
+
+std::vector<CChainState*> ChainstateManager::GetAll()
+{
+    std::vector<CChainState*> out;
+
+    if (!IsSnapshotValidated() && m_ibd_chainstate) {
+        out.push_back(m_ibd_chainstate.get());
+    }
+
+    if (m_snapshot_chainstate) {
+        out.push_back(m_snapshot_chainstate.get());
+    }
+
+    return out;
+}
+
+CChainState& ChainstateManager::InitializeChainstate(
+    bool activate, const uint256& snapshot_blockhash)
+{
+    std::unique_ptr<CChainState>& to_modify = (
+        snapshot_blockhash.IsNull() ? m_ibd_chainstate : m_snapshot_chainstate);
+
+    to_modify.reset(new CChainState(snapshot_blockhash));
+
+    if (activate) {
+        LogPrintf("Switching active chainstate to %s\n", to_modify->ToString());
+        m_active_chainstate = to_modify.get();
```

Comment:
in commit 16d20d91b43b1c30333fe14272d69cfb543229b3:

You mention that "// If a snapshot chainstate exists, it will always be our active."

However, then you provide code here to activate a ibd chainstate even though a snapshot chainstate might exist. This is confusing and the code should probably not provide such a backdoor.

Reply:
Maybe a check like `if (m_active_chainstate && to_modify != m_snapshot_chainstate) LogPrintf("Error!"); else ...` or similar could prevent that?

Reply:
Yep, this is a really good point. Fixed it to only allow a single activation sequence.